### PR TITLE
Generate proper API description for Map types using "additionalProperties"

### PR DESCRIPTION
**Pull request by clementdenis**
This PR changes the way Cloud Endpoints handles Map types when generating API specs:
- Previously, maps were always described as JsonMap, making the API Explorer buggy, and the Java clients barely usable.
- The new behavior is to use "additionalProperties" in the JsonSchema, that work really well for String-keyed maps, and generate correctly typed Java clients.
- A flag system based both on env variables and system properties (see MapSchemaFlag) can be used to fine tune the new behavior.

**This introduces backward incompatible changes to the generated Java clients, and the stricter validation of acceptable Map types can make some APIs fail to generate the spec.**
Legacy behavior can be reintroduced for both of these using flags in MapSchemaFlag (see FORCE_JSON_MAP_SCHEMA and IGNORE_UNSUPPORTED_KEY_TYPES).

A note about Map types with array-like values:
- This use case is properly supported by both the Discovery and Swagger formats, and works fine in the API explorer.
- However, it is not supported by the [cloud-based client library generator](https://google-api-client-libraries.appspot.com/generate) (which is presumably base on [this project](https://github.com/google/apis-client-generator)), it will generate invalid Java code.
- So the default behavior is to keep using JsonMap schema for the array-valued maps, but it can be forced to use additionalProperties with the MapSchemaFlag.SUPPORT_ARRAYS_VALUES flag.

I previously opened an issue about this: https://github.com/cloudendpoints/endpoints-java/issues/142

### CODE REVIEW

**Comment by tangiel** [comment]
Looks like there is a conflict now (may have merged out of order)

**Comment by clementdenis** [comment]
I rebased from latest master (conflict was only in imports).

File: endpoints-framework/src/test/java/com/google/api/server/spi/discovery/DiscoveryGeneratorTest.java [hunk 1]
```diff
@@ -110,6 +116,57 @@ public void testWriteDiscovery_ArrayEndpoint() throws Exception {
...
+      RestDescription doc = getDiscovery(new DiscoveryContext(), MapEndpoint.class);
+      RestDescription expected = readExpectedAsDiscovery("map_endpoint_legacy.json");
+      compareDiscovery(expected, doc);
+    } finally {
+      System.clearProperty(FORCE_JSON_MAP_SCHEMA.systemPropertyName);
+    }
+  }
+
+  @Test
+  public void testWriteDiscovery_MapEndpoint_InvalidKeyTypeKO() throws Exception {
```
Comment (tangiel) [inline endpoints-framework/src/test/java/com/google/api/server/spi/discovery/DiscoveryGeneratorTest.java]: Is "KO" intentional?

File: test-utils/src/main/java/com/google/api/server/spi/testing/MapEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (tangiel) [inline test-utils/src/main/java/com/google/api/server/spi/testing/MapEndpoint.java]: Let's keep the year up to date on new files.

File: endpoints-framework/src/test/java/com/google/api/server/spi/discovery/DiscoveryGeneratorTest.java (hunk 1 above)
Comment (clementdenis) [reply endpoints-framework/src/test/java/com/google/api/server/spi/discovery/DiscoveryGeneratorTest.java]: It is, but I will make the name more explicit.

**Comment by clementdenis** [comment]
I fixed formatting and copyright issues.

**Pull request merged**